direct/core/src/main/java/cz/o2/proxima/direct/batch/BatchLogReaders.java

```diff
@@ -0,0 +1,184 @@
+
+/** Class constructing various {@link BatchLogReader BatchLogReaders}. */
+public class BatchLogReaders {
+
+  /**
+   * Create throughput limited {@link BatchLogReader}.
+   *
+   * @param delegate delegate to read data from
+   * @param limiter the throughput limiter
+   * @return throughput limited {@link BatchLogReader}
+   */
+  public static BatchLogReader withLimitedThroughput(
+      BatchLogReader delegate, ThroughputLimiter limiter) {
+    return new ThroughputLimitedBatchLogReader(delegate, limiter);
+  }
+
+  private static class ForwardingLimitedBatchLogReader implements BatchLogReader {
+
+    @Delegate private final BatchLogReader delegate;
+
+    private ForwardingLimitedBatchLogReader(BatchLogReader delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override
+    public String toString() {
+      return delegate.toString();
+    }
+  }
+
+  private static class ThroughputLimitedBatchLogReader extends ForwardingLimitedBatchLogReader {
+
+    private final ThroughputLimiter limiter;
+
+    public ThroughputLimitedBatchLogReader(BatchLogReader delegate, ThroughputLimiter limiter) {
+      super(delegate);
+      this.limiter = limiter;
+    }
+
+    @Override
+    public void observe(
+        List<Partition> partitions,
+        List<AttributeDescriptor<?>> attributes,
+        BatchLogObserver observer) {
+
+      super.observe(partitions, attributes, throughputLimited(observer, partitions));
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("limiter", limiter)
+          .add("delegate", super.toString())
+          .toString();
+    }
+
+    private BatchLogObserver throughputLimited(
+        BatchLogObserver delegate, List<Partition> consumedPartitions) {
+
+      return new ThroughputLimitedBatchLogObserver(
+          delegate, getPartitions().size(), consumedPartitions, limiter);
+    }
+  }
+
+  private static class ForwardingBatchLogObserver implements BatchLogObserver {
+
+    @Delegate private final BatchLogObserver delegate;
+
+    private ForwardingBatchLogObserver(BatchLogObserver delegate) {
+      this.delegate = delegate;
+    }
+  }
+
+  private static class ThroughputLimitedBatchLogObserver extends ForwardingBatchLogObserver {
+
+    private final int numPartitions;
+    private final Collection<Partition> assignedPartitions;
+    private final ThroughputLimiter limiter;
+    private long watermark = Long.MIN_VALUE;
+
+    public ThroughputLimitedBatchLogObserver(
+        BatchLogObserver delegate,
+        int numPartitions,
+        Collection<Partition> assignedPartitions,
+        ThroughputLimiter limiter) {
+
+      super(delegate);
+      this.numPartitions = numPartitions;
+      this.assignedPartitions = new ArrayList<>(assignedPartitions);
+      this.limiter = Objects.requireNonNull(limiter);
+    }
+
+    @Override
+    public boolean onNext(StreamElement element) {
+      if (throwsInterrupted(this::waitIfNecessary)) {
```

Comment:
Not sure about threading model here. If every reader has dedicated thread, sleeping here should be ok.